src/core/thread/mlr_manager.cpp

```diff
@@ -81,6 +83,126 @@ void MlrManager::HandleBackboneRouterPrimaryUpdate(BackboneRouter::Leader::State
     UpdateReregistrationDelay(needRereg);
 }
 
+#if OPENTHREAD_CONFIG_MLR_ENABLE
+void MlrManager::UpdateLocalSubscriptions(void)
+{
+#if OPENTHREAD_CONFIG_TMF_PROXY_MLR_ENABLE
+    // Check multicast addresses are newly listened against Children
+    for (Ip6::ExternalNetifMulticastAddress &addr : Get<ThreadNetif>().IterateExternalMulticastAddresses())
+    {
+        if (addr.GetAddress().IsMulticastLargerThanRealmLocal() && addr.GetMlrState() == kMlrStateToRegister &&
+            IsAddressMlrRegisteredByAnyChild(addr.GetAddress()))
+        {
+            addr.SetMlrState(kMlrStateRegistered);
+        }
+    }
+#endif // OPENTHREAD_CONFIG_TMF_PROXY_MLR_ENABLE
+
+    ScheduleSend(0);
+}
+
+bool MlrManager::IsAddressMlrRegisteredByNetif(const Ip6::Address &aAddress)
+{
+    bool ret = false;
+
+    for (const Ip6::ExternalNetifMulticastAddress &addr : Get<ThreadNetif>().IterateExternalMulticastAddresses())
+    {
+        if (addr.GetAddress().IsMulticastLargerThanRealmLocal() && addr.GetMlrState() == kMlrStateRegistered &&
+            addr.GetAddress() == aAddress)
+        {
+            ExitNow(ret = true);
+        }
+    }
+
+exit:
+    return ret;
+}
+#endif // OPENTHREAD_CONFIG_MLR_ENABLE
+
+#if OPENTHREAD_CONFIG_TMF_PROXY_MLR_ENABLE
+
+bool MlrManager::IsAddressMlrRegisteredByAnyChildExcept(const Ip6::Address &aAddress, const Child *aExceptChild)
+{
+    bool ret = false;
+
+    for (Child &child : Get<ChildTable>().Iterate(Child::kInStateValidOrRestoring))
```

Comment:
Any thoughts on whether or not we should consider a restoring Child?

Reply:
When the parent is up after reboot, it will set its children as [kStateRestored](https://github.com/openthread/openthread/blob/master/src/core/thread/mle_router.cpp#L3688); after it re-establish link with neighboring routers, it will try to restore its children via ChildUpdateRequest() and set the state as [kStateChildUpdateRequest](https://github.com/openthread/openthread/blob/master/src/core/thread/mle_router.cpp#L3100).

The parent won't store the MAs for its children,  it only knows the MLEID of the child before restore the link with its child.
It would be useless to iterate the children in restoring state.

thoughts?

Reply:
I digged into the code for a little bit:

**`UpdateChildAddresses` is called in 3 places:**

1. `HandleChildIdRequest` which can change Child state from `Invalid` to `ChildIdRequest` or 
2. `HandleChildUpdateRequest` which can change Child state from `Restoring` to `Valid`
3. `HandleChildUpdateResponse` which can change Child state from `ChildUpdateRequest` to `Valid`

One thing I noticed is that in both case 2 and 3, `UpdateChildAddresses` is called before `SetChildStateToValid`. So, I think it necessary to also consider `Restoring` state since it just going to be `Valid` immediately. 

For case 1 (`HandleChildIdRequest`),  `UpdateChildAddresses` seems to be called when the Child state is still invalid. In such case, no address will be processed by `MlrManager`, which I think is an issue. 

Thoughts? @librasungirl

Reply:
Generally speaking, I think only the keep-alive child update request could for sure include all unicast/multicast that should be registered.

For case 3,  ChildUpdateResponse from child only include MLEID.
https://github.com/openthread/openthread/blob/c5e7252fed253760e76097179a1078f410bcd641/src/core/thread/mle.cpp#L2324

For case 2, when the child is `Restoring`, if it is a fragmented keep-alive CUR from child, it may have no chance to go up to MLE layer.  
If tracing back to #1853 which introduced this change, it is a Child Update Request from child which reboots, which in theory should be better to only include mesh local eid (#5295 is submitted for this).

Thoughts?

Reply:
> For case 3, ChildUpdateResponse from child only include MLEID.
We should also consider other implementations (e.g. Kirale)? We can't assure that other implementations have same behavior as OpenThread. 

> For case 2, when the child is Restoring, if it is a fragmented keep-alive CUR from child, it may have no chance to go up to MLE layer.

I am not clear about the background. The CUR must be be fragmented in such case?

Moreover, what do you have in mind for case 1? I think we can't assume that `CIR` contains only MLEID.

Reply:
Thread requires fragmented MLE messages to enable MAC frame security. However, if the receiving device does not yet have the necessary MAC frame counters (e.g. after reboot), then fragmented MLE messages are dropped.

Implicitly, this means any message that includes a Challenge and/or Response TLV must be small enough to fit within a single IEEE 802.15.4 frame. At the same time, if an Address Registration TLV with multiple IPv6 addresses can fit within a single IEEE 802.15.4 frame, then it can carry more than just the ML-EID.

Reply:
@simonlingoogle 
For case 1, the filter `kInStateValidOrRestoring` itself does already exclude child in `kStateParentResponse` (before receiving CIR) or `kStateChildIdRequest` (after receiving CIR), right?

Reply:
> @simonlingoogle
> For case 1, the filter `kInStateValidOrRestoring` itself does already exclude child in `kStateParentResponse` (before receiving CIR) or `kStateChildIdRequest` (after receiving CIR), right?

I believe the Child is mostly likely in state `kStateParentRequest` when the Parent receives a `Child Id Request` (I couldn't find Parent set Child state to `kStateParentResponse` anywhere in the code). The Parent will call `UpdateChildAddresses` in the  `kStateParentRequest` state before transferring it to `kStateChildIdRequest`. You are right that both states are filtered-out by `IsStateValidOrRestoring`. I think this is a potential issue that we need to fix.

Reply:
Analyzed a Child's state transition: 
![image](https://user-images.githubusercontent.com/52448427/88484550-fc6e6900-cfa1-11ea-9c3b-ae7c50da1252.png)

I believe that the filtered state actually does not matters that much, whether it's `Valid` only or `ValidOrRestoring` . What we need to do is two things:

- When the Child enters the filtered state (`Valid` or `ValidOrRestoring`), we need the check the current multicast addresses and schedule MLR.req if necessary.
- When the Child sends Address Registration TLV (by ChildUpdateRequest or ChildUpdateResponse) in the filtered state, we update the multicast addresses and schedule MLR.req if necessary. 

So, if we filter `Valid` state, then we need to update multicast addresses when a Child's state becomes `kStateValid`. Then, only update multicast addresses in `kStateValid` state. 

If we filter `ValidOrRestoring` state, then we need to update multicast addresses when a Child's state becomes `kStateValid` or `kStateChildUpdateRequest`. Then, update multicast addresses in these two states.  

With this analysis, maybe filtering `Valid` state is a better idea because we already have a clear entrance for a Child to become valid: `MleRouter::SetChildStateToValid` which calls `Signal(OT_NEIGHBOR_TABLE_EVENT_CHILD_ADDED, aChild);`. We can update the initial multicast addresses in these two `SetChildStateToValid` or `Signal`. 

Thoughts? @librasungirl

Reply:
Revised accordingly in the last commit. @librasungirl

Reply:
Thanks for your excellent efforts for this minor enhancement. SGTM.